### PR TITLE
[`flake8-simplify`] Stabilize detection of Yoda conditions for "constant" collections (`SIM300`)

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

See: https://github.com/astral-sh/ruff/pull/9164.

We made a bunch of improvements to the rule, to handle "constant" collections (e.g., lists that consist of constants). This will lead to new violations, but I do think it's a better behavior and it's clearly stable.

### CODE REVIEW

**Comment by charliermarsh** [comment]
A bit more controversial than the others because the scope is larger (hopefully ecosystem reflects that).

File: crates/ruff_linter/src/rules/flake8_simplify/rules/yoda_conditions.rs [hunk 1]
```diff
@@ -96,7 +95,7 @@ enum ConstantLikelihood {
 
 impl ConstantLikelihood {
     /// Determine the [`ConstantLikelihood`] of an expression.
-    fn from_expression(expr: &Expr, preview: PreviewMode) -> Self {
+    fn from_expression(expr: &Expr) -> Self {
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/flake8_simplify/rules/yoda_conditions.rs:98]: Should this now be `impl From<&Expr> for ConstantLikelihood {}`? (https://github.com/astral-sh/ruff/pull/9164#discussion_r1433002869)

**Comment by AlexWaygood** [comment]
I pushed to your branch to fix some trivial clippy issues, but looks like some snapshots are out of date as well

**Review by AlexWaygood — APPROVED** [review]
Lots of ecosystem hits, but that's why it's a minor release! Agreed that this is a big improvement

**Comment by MichaReiser** [comment]
It seems that the vast majority of new violations is for usages in tests, specifically inside of `assert`. I think the rule should be disabled for assertions because a common pattern is to write the expected value on the left, which often is a literal. 

I'm also not sure if I would count this as a yoga condition:

```
assert [] == get_removed_provider_ids()
```

The right isn't a variable, it's a function call.

**Comment by AlexWaygood** [comment]
I would always put the thing I was testing (the result of the function call) on the left, and the thing I expected it to be on the right. `assert [] == my_function_call()` is just as much a Yoda condition as anything else this rule is detecting, in my opinion.

**Comment by AlexWaygood** [comment]
I'm pretty strongly of the opinion that this is a clear improvement for this rule. In all of the Python codebases I've worked on, whether the test suite is using `unittest` or `pytest`, the object being tested goes on the left-hand side of the assertion. If some codebases have the opposite convention, they may just want to switch this rule off for their codebases, but in my opinion this change makes the rule more consistent and smarter without changing the basic premise of the rule. If you disagree with the premise, you can always switch it off ;-)

Pytest doesn't appear to display it any better or worse depending on the order of the assertion:

image

**Pull request merged**